hail/src/main/scala/is/hail/expr/ir/PruneDeadFields.scala

```diff
@@ -1072,7 +1072,7 @@ object PruneDeadFields {
       if (upcastEntries && mt.typ.entryType != rType.entryType)
         mt = MatrixMapEntries(mt, upcast(Ref("g", mt.typ.entryType), rType.entryType))
 
-      if (upcastRows && mt.typ.rowType != rType.rowType)
+      if (upcastRows && mt.typ.rvRowType != rType.rvRowType)
```

Comment:
I don't think this actually fixes the underlying problem. The issue is that if you have:

``` { A, [the entries: B], C} ``` and want ```{ [the entries: b], c}``` where `a` is a supertype of `A` and so on, then the MatrixMapEntries upcast will put the entries at the end, making the types incompatible.

I don't think I'm explaining this well.

One option may be to implement this cast using the new IR conversion stuff Patrick has just PRed, because upcasting a table doesn't have this issue.

Reply:
Why would the matrix map entries upcast put the entries at the end?

Reply:
ahhhh wait, you're right -- it inserts in the same place, right? We haven't made that change yet.

Reply:
but then I'm not understanding how this fixes the problem. We've just upcast the entries in the lines above, so aren't these checks equivalent?

Reply:
If an optimization puts the entries in the wrong place, then the upcast of the entries would keep it there (in the wrong place). But maybe we should be figuring out why that would happen?

Reply:
but if the entries are in the wrong place, an upcast can't possibly fix it -- the the supertype you're trying to upcast to is *not* a supertype of the out-of-order thing

Reply:
@chrisvittal does this test you've added fail without this code change?

Reply:
yeah, this fix kind of stretches the definition of "upcast"---it becomes more of a "change_type" thing.

Reply:
yes

Reply:
huh, I don't get it.

I'm also fine with making this change while we poke around further. It seems pretty safe to me.

Reply:
if we're doing something weird we'll hit that assertion error in MAtrixMapRows that Amanda had the good idea to add.

Reply:
@tpoterba can I approve this PR, then?

Reply:
sure, but I want to know why this fixes it

Reply:
to make myself clearer, I'm happy to merge this but I also want to poke around in the next few days to figure out why this works.